File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/feature/product/entity/Categories.java
package com.multicampus.gamesungcoding.a11ymarketserver.feature.product.entity;

import com.multicampus.gamesungcoding.a11ymarketserver.common.id.UuidV7;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.util.List;
import java.util.UUID;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Categories {
    @Id
    @UuidV7
    @Column(length = 16, nullable = false, updatable = false)
    private UUID categoryId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_cat_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    private Categories parentCategory;

    @Column(length = 200, nullable = false)
    private String categoryName;

    @OneToMany(mappedBy = "category")
    private List<Product> products;

    @Builder
    private Categories(Categories parentCategory,
                       String categoryName) {
        
        this.parentCategory = parentCategory;
        this.categoryName = categoryName;
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/feature/product/entity/Categories.kt
package com.multicampus.gamesungcoding.a11ymarketserver.feature.product.entity

import com.multicampus.gamesungcoding.a11ymarketserver.common.id.UuidV7
import jakarta.persistence.*
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.util.*

@Entity
@Table(name = "categories")
@EntityListeners(AuditingEntityListener::class)
class Categories(
    @Column(length = 200, nullable = false)
    val categoryName: String,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_cat_id")
    @OnDelete(action = OnDeleteAction.CASCADE)
    val parentCategory: Categories? = null
) {
    @Id
    @UuidV7
    @Column(length = 16, nullable = false, updatable = false)
    var categoryId: UUID? = null
        private set

    @OneToMany(mappedBy = "category")
    var products: MutableList<Product> = mutableListOf()
        private set
}
